changed ast to make conjunction a list of goals, modified parser and functions common accordingly

ZaidQureshi opened the pull request:
This makes a fundamental change to our ast. Now, a Clause has a head and a body where the body is a list of exps. The list represents a conjunction between all the exps in the list. Similarly a Query is a list of exps where the lsit represents a conjunction between all the exps in the list. This is to simplify the evaluation. Disjunction has been removed. I didn't modify any of the test files.

adamjstewart commented on src/ast.ml:
```diff
@@ -10,13 +10,13 @@ type exp =
-    | ConjunctionExp of exp * exp   (* term, term               *)
-    | DisjunctionExp of exp * exp   (* term; term               *)
+    (*  | ConjunctionExp of exp * exp   (* term, term               *)*)
+(* | DisjunctionExp of exp * exp   (* term; term               *)*)
```
If you're going to get rid of these, delete them, don't comment them out.

adamjstewart commented on src/common.ml:
```diff
@@ -47,15 +47,19 @@ let rec string_of_exp e =
         "ConjunctionExp (" ^ (string_of_exp e1) ^ ", " ^ (string_of_exp e2) ^ ")"
     | DisjunctionExp (e1, e2) ->
         "DisjunctionExp (" ^ (string_of_exp e1) ^ ", " ^ (string_of_exp e2) ^ ")"
+ *)
```
Same here.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,26 @@
+    | ConjunctionExp(e1,e2) -> (match eval_query (lift_sub_goal e1 env) with
+                               | None -> None
+                               | Some(s1) -> eval_query (lift_sub_goal (lift_sub_goal e2 env) s1)
+ *)
```
Same here.

adamjstewart commented on src/parser.mly:
```diff
@@ -66,9 +66,9 @@ clause:
-    | pl = predicate_list; SEMICOLON; p = predicate     { DisjunctionExp (pl, p) }
+    | p = predicate                                     { [p] }
+    | p = predicate; COMMA; pl = predicate_list        { p :: pl }
+ (*   | pl = predicate_list; SEMICOLON; p = predicate     { DisjunctionExp (pl, p) }*)
```
Same here.

adamjstewart commented on src/common.ml:
```diff
@@ -47,21 +47,28 @@ let rec string_of_exp e =
+let string_of_atom a =
+    match a with
+    | (TermExp(s,_)) -> s
+    | _ -> raise(Failure "not needed")
```
I would rename this `functor_of_termexp` and require the input to be a `TermExp` only.

ZaidQureshi replied:
Restricting the type that can be passed into string_of_atom causes warnings. In fact to truly solve this issue we would need to modify our AST slightly but that isn't necessary.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,134 @@
 open Ast
+open Common
```
I would throw in 2 empty lines to separate imports from code.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,134 @@
+                                                          | ConstExp(c) -> (find_vars xs)
+                                                          | TermExp(s,el) -> (find_vars el) @ (find_vars xs))
+                                            
+(* from https://rosettacode.org/wiki/Remove_duplicate_elements#OCaml *)
```
OCaml does support sets: https://caml.inria.fr/pub/docs/manual-ocaml/libref/Set.html

I wonder if it would be simpler to use them.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,134 @@
+let rec occurs s t = match s with VarExp(n) -> (match t with VarExp(m) -> n = m
+                                                          | TermExp(st,el) -> List.fold_left (fun acc v -> acc || (occurs s v)) false el
+                                                          | _ -> false)
+                                |  _ -> raise(Failure "not needed")
```
You should put a single empty line between each function definition.

adamjstewart commented on src/parser.mly:
```diff
@@ -54,8 +54,8 @@
 
```
If we're going to get rid of all references to Disjunctions remember to remove the SEMICOLON token above.

ZaidQureshi commented:
Fixed all of your tests for the modifications of the ast. Also removed the semicolon token from lexer, parser, and their tests. Regarding stuff from evaluator.ml, things are only "easier" if nothing is implemented and things don't work (which isn't our case).  Also, restricting the type that  can be passed into string_of_atom causes warnings. In fact to truly solve this issue we would need to modify our AST slightly but that isn't necessary.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,146 @@
-        then true
-        else check_if_dec_in_db (Clause (h,b) ,xs)
+let (fresh, reset) =
+  let nxt = ref 0 in
```
What's a `ref` and what does `:=` do?

ZaidQureshi replied:
ref is a makes it a reference type (like pointers).

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,146 @@
+                   
+let rec find_vars q  =
+  match q with [] -> []
+             | (x::xs) -> (match x with VarExp(v) -> [x] @ (find_vars xs)
```
You could just use `x :: (find_vars xs)` here.

ZaidQureshi replied:
yes I can.

adamjstewart commented on src/evaluator.ml:
```diff
@@ -1,31 +1,146 @@
+  let rec tail_uniq a l =
+    match l with
+    | [] -> a
+    | hd::tl -> tail_uniq (hd::a) (List.filter (fun x -> x  <> hd) tl) in
```
What is `<>`?

ZaidQureshi replied:
inequality test

adamjstewart commented:
Could you add docstrings to every function you wrote? That would help greatly in understanding what it is supposed to do.

ZaidQureshi commented:
Yes, also the report should help. If/when you are back in town maybe we should meet up so I can explain it.
